### PR TITLE
Ensemble Models

### PR DESCRIPTION
**Pull request by aditya0by0**
- Has Dependency with PR #64

### CODE REVIEW

**Comment by aditya0by0** [comment]
Test example config : 

```bash
class_path: chebai.models.ensemble.ChebiEnsemble
init_args:
  optimizer_kwargs:
    lr: 1e-3
  model_configs: {
    "chebi50": {
      "path": "logs/chebi50_bce_unweighted/version_9/checkpoints/best_epoch=00_val_loss=0.3521_val_macro-f1=0.0381_val_micro-f1=0.1753.ckpt",
      "TPV": 0.7,
      "FPV": 0.9
    },

    "chebi50_1": {
      "path": "logs/chebi50_bce_unweighted/version_9/checkpoints/best_epoch=00_val_loss=0.3521_val_macro-f1=0.0381_val_micro-f1=0.1753.ckpt",
      "TPV": 0.8,
      "FPV": 0.7
    },
  }

```

File: configs/model/ensemble/chebiEnsemble.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class_path: chebai.models.ensemble.ChebiEnsemble
+init_args:
+  optimizer_kwargs:
+    lr: 1e-3
```
Comment (sfluegel05) [inline configs/model/ensemble/chebiEnsemble.yml:4]: Why does the ensemble model need a learning rate? I would assume that it cannot be trained at all and is a prediction-only model

Comment (aditya0by0) [reply configs/model/ensemble/chebiEnsemble.yml]: Thanks for pointing it out. I have removed the learning rate part from the config.

**Comment by aditya0by0** [comment]
Please find the latest config file eg... which is working for electra and gnn 
```bash
class_path: chebai.ensemble.FullEnsembleWMV
init_args:
  data_processed_dir_main: "data/chebi_v231/ChEBI50/processed"
  operation_mode: "prediction"
  smiles_list_file_path: "data/chebi_v231/ChEBI50/processed/smiles.txt"
  _perform_validation_checks: False  # To avoid check for using same model with same model configs
  model_configs: {
    "chebi50": {
      "model_ckpt_path": "logs\\chebi50_electra_231\\version_1\\checkpoints\\best_epoch=00_val_loss=0.6766_val_macro-f1=0.7273_val_micro-f1=0.0248.ckpt",
      "model_config_file_path": "configs\\model\\electra.yml",
      "data_config_file_path": "configs\\data\\chebi\\chebi50.yml",
      "model_labels_path": "data/chebi_v231/ChEBI50/processed/classes.json",
      "wrapper_class_path": chebai.ensemble.NNWrapper,
    },

    "chebi50_1": {
      "model_ckpt_path": "logs\\chebi50_electra_231\\version_1\\checkpoints\\best_epoch=00_val_loss=0.6766_val_macro-f1=0.7273_val_micro-f1=0.0248.ckpt",
      "model_config_file_path": "configs\\model\\electra.yml",
      "data_config_file_path": "configs\\data\\chebi\\chebi50.yml",
      "model_labels_path": "data/chebi_v231/ChEBI50/processed/classes.json",
      "wrapper_class_path": chebai.ensemble.NNWrapper,
    },

    "chebi50_graph": {
      "model_ckpt_path": "logs\\regated_gnn_v231\\version_0\\checkpoints\\best_epoch=00_val_loss=0.6786_val_macro-f1=0.6364_val_micro-f1=0.0321.ckpt",
      "model_config_file_path": "..\\python-chebai-graph\\configs\\model\\gnn_res_gated.yml",
      "data_config_file_path":  "..\\python-chebai-graph\\configs\\data\\chebi50_graph_properties.yml",
      "model_labels_path": "data/chebi_v231/ChEBI50/processed/classes.json",
      "wrapper_class_path": chebai.ensemble.GNNWrapper,
    },

    # "chemlog": {
    #   "model_labels_path": "data/chebi_v231/ChEBI50/processed/classes.json",
    #   "wrapper_class_path": chebai.ensemble.ChemLogWrapper,
    # },
  }

```

File: chebai/ensemble/_wrappers/_chemlog.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    def _predict_from_list_of_smiles(self, smiles_list):
+        return {"logits": self.get_chemlog_results(smiles_list)}
+
+    def _evaluate_from_data_file(self, **kwargs) -> list:
+        data_pkl_file_path = kwargs["data_pkl_file_path"]
+        data_df = pd.read_pickle(data_pkl_file_path)
+        smiles_list = data_df["SMILES"].to_list()
+        return {"logits": self.get_chemlog_results(smiles_list)}
+
+    def get_chemlog_results(self, smiles_list) -> list:
```
Comment (aditya0by0) [inline chebai/ensemble/_wrappers/_chemlog.py:30]: @sfluegel05, I have added wrapper for chemlog. What I need from the method is some kind of prediction for each class, which can be passed to sigmoid function later on.

I am not sure how chemlog achieves this. Can you please guide me?
